graylog2-server/src/main/java/org/graylog2/database/entities/ScopedEntityPaginatedDbService.java

```diff
@@ -0,0 +1,102 @@
+
+import org.graylog2.bindings.providers.MongoJackObjectMapperProvider;
+import org.graylog2.database.MongoConnection;
+import org.graylog2.database.PaginatedDbService;
+
+import java.util.Objects;
+import java.util.Optional;
+
+/**
+ * A base database service to handle persistence and deletion of {@link ScopedEntity} instance.  Persistence and deletion is performed by the parent class, {@link PaginatedDbService},  this service simply performs mutability checks.
+ *
+ * <p>
+ * A {@link EntityScopeService} is used to perform the actual mutability checks based on the entity's <b>scope</b>.
+ * </p>
+ *
+ * @param <E> type parameter for a {@link ScopedEntity}'s subclass.
+ */
+public abstract class ScopedEntityPaginatedDbService<E extends ScopedEntity> extends PaginatedDbService<E> {
+
+    protected final EntityScopeService entityScopeService;
+
+    public ScopedEntityPaginatedDbService(MongoConnection mongoConnection,
+                                          MongoJackObjectMapperProvider mapper,
+                                          Class<E> dtoClass, String collectionName,
+                                          EntityScopeService entityScopeService) {
+        super(mongoConnection, mapper, dtoClass, collectionName);
+        this.entityScopeService = entityScopeService;
+    }
+
+    public final boolean isMutable(ScopedEntity scopedEntity) {
+
+        Objects.requireNonNull(scopedEntity, "Entity must not be null");
+
+        // First, check whether this entity has been persisted, if so, the persisted entity's scope takes precedence.
+        Optional<E> current = get(scopedEntity.id());
+        if (current.isPresent()) {
+            return entityScopeService.isMutable(current.get());
+        }
+
+        // The entity does not exist in the database, This could be a new entity--check it
+        return entityScopeService.isMutable(scopedEntity);
+    }
+
+    /**
+     * Find a Mutable entity with the given ID.  <b>empty</b> is returned if the entity is not found OR if it is <b>immutable</b>.
+     *
+     * @param id target ID
+     * @return optional target entity
+     */
+    public final Optional<E> findMutableById(String id) {
+
+        return get(id)
+                .filter(entityScopeService::isMutable);
+    }
+
+    @Override
+    public E save(E entity) {
+
+        ensureValidScope(entity);
+        if (entity.id() != null) {
+            ensureMutability(entity);
+        }
+        return super.save(entity);
+    }
+
+    public final void ensureValidScope(E entity) {
+        if (!entityScopeService.hasValidScope(entity)) {
+            throw new IllegalArgumentException("Invalid Entity Scope: " + entity.scope());
+        }
+    }
+
+
+    public final void ensureMutability(E entity) {
+        if (!isMutable(entity)) {
+            throw new IllegalArgumentException("Immutable entity cannot be modified");
+        }
+    }
+
+    @Override
+    public int delete(String id) {
```

Comment:
Should the `#delete` method also be `final` to avoid unwanted mutation in subclasses?

Reply:
I think so. What do you think @roberto-graylog?